src/Tools/ExternalAccess/LiveShare/Classification/RoslynSyntaxClassificationService.cs

```diff
@@ -0,0 +1,192 @@
+                    _originalService.AddLexicalClassifications(text, textSpan, result, cancellationToken);
+                }
+            }
+            else if (_experimentationService.IsExperimentEnabled(WellKnownExperimentNames.SyntacticExp_Remote))
+            {
+                var documentId = _remoteLanguageServiceWorkspace.GetDocumentIdInCurrentContext(text.Container);
+                var document = _remoteLanguageServiceWorkspace.CurrentSolution.GetDocument(documentId);
+                if (document == null)
+                {
+                    // It is expected that the document cannot be found (e.g. source text comes from preview workspace).
+                    // These cases are already not supported in liveshare, so for now just return.
+                    return;
+                }
+                using (new RequestLatencyTracker(SyntacticLspLogger.RequestType.LexicalClassifications))
+                {
+                    _threadingContext.JoinableTaskFactory.Run(async () =>
+                    {
+                        await AddRemoteClassificationsAsync(LexicalClassificationsHandler.LexicalClassificationsMethodName, document.FilePath, text, textSpan, result, cancellationToken).ConfigureAwait(false);
+                    });
+                }
+            }
+            else
+            {
+                // Some other invalid flight.  Just fallback to the regular service.  Don't want to block the user based on an experimentation failure.
+                _originalService.AddLexicalClassifications(text, textSpan, result, cancellationToken);
+            }
+        }
+
+        public void AddSemanticClassifications(SemanticModel semanticModel, TextSpan textSpan, Workspace workspace, Func<SyntaxNode, ImmutableArray<ISyntaxClassifier>> getNodeClassifiers, Func<SyntaxToken, ImmutableArray<ISyntaxClassifier>> getTokenClassifiers, ArrayBuilder<ClassifiedSpan> result, CancellationToken cancellationToken)
+        {
+            _threadingContext.JoinableTaskFactory.Run(async () =>
+            {
+                var sourceText = await semanticModel.SyntaxTree.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                await AddRemoteClassificationsAsync(RoslynMethods.ClassificationsName, semanticModel.SyntaxTree.FilePath, sourceText, textSpan, result, cancellationToken).ConfigureAwait(false);
+            });
+        }
+
+        public async Task AddSemanticClassificationsAsync(Document document, TextSpan textSpan, Func<SyntaxNode, ImmutableArray<ISyntaxClassifier>> getNodeClassifiers, Func<SyntaxToken, ImmutableArray<ISyntaxClassifier>> getTokenClassifiers, ArrayBuilder<ClassifiedSpan> result, CancellationToken cancellationToken)
+        {
+            // If we are in the preview workspace, the request is to try to colorize a lightbulb preview. The document would have changed
+            // in this workspace and we currently don't support requests for anything but the Workspace.CurrentSolution.
```

Comment:
wait... why not?  the remote process should have no problem forking workspaces.  That's how we designed OOP to work.

Reply:
this is not OOP. this is liveshare.

Reply:
Are there any docs on how liveshare works so i can understand and comment on the implications of these decisions and the areas teh team should be looking into here?

Reply:
> 
> 
> Are there any docs on how liveshare works so i can understand and comment on the implications of these decisions and the areas teh team should be looking into here?

I believe the best available are these - https://docs.microsoft.com/en-us/visualstudio/liveshare/
and these - https://www.nuget.org/packages/Microsoft.VisualStudio.LiveShare/#
@daytonellwanger might have more.

Reply:
@daytonellwanger Are there any docs on the low level design of liveshare?  specifically, i'm looking to understand how it works from a client/server perspective.  Where is information expected to live?  Who is responsible for "source of truth".  How clients and servers synchronize and ask questions in a way that is consistent, etc. etc.

For example, here's an easy sort of problem i can envision, which i'd like an undertanding of how it is handled:

1. on the client (without roslyn), user makes some edits.  this increases the length of the file.
2. client then wants to classify the text at the end of the file.
3. tehy then communicate with server, but server doesn't have text/project that matches what client has.  server passes requests onto roslyn, which crashes because invariant is broken.

Now, given this is such a simple scenario, and you're likkely not crashing.  You have presumably solved this in some manner.  For example, some sort of syncing.  Or perhaps validating that the request can even be asked on the other end.  However, those designs and invariants are not clear here, and as such it makes it *very* hard to tell if the code here is correct.

--

To give some context, i was involved in the early days of the LSP with the TS lang server.  There were all sorts of assumptions and terrible practices going on between the clients and the servers, and at the end of the day, it became necessary for tons of hacks to be added all over the place to just paper over these places where little discussion/design had happened, and people jsut did random stuff because it seemed to work without understanding the implications of those decisions.

I would *very very* much like to avoid that in roslyn.  For a very large part of it we've operated under the assumption that it's much better to have strict invariatns and to fail early to ensure that people "do the right thing".  If we're still abiding by that, that's great.  But if we're not, it's important to document so we cna ensure everyone's mental model of responsibilities is accurate.

Thanks!

Reply:
There's an engineering talk about Live Share in Visual Studio available [here](https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki?wikiVersion=GBwikiMaster&pagePath=%2FTeam%20Pages%2FVS%20IDE%2FEngineering%20Talks&pageId=1075).

Reply:
That is not helpful to me :D